pkg/realmd/operation.js

```diff
@@ -320,6 +325,77 @@
             return creds;
         }
 
+        // Request and install a kerberos keytab for cockpit-ws (with IPA)
+        // This is opportunistic: Some realms might not use IPA, or an unsupported auth mechanism
+        function install_ws_keytab() {
+            // skip this on remote ssh hosts, only set up ws hosts
+            if (cockpit.transport.host !== "localhost")
+                return cockpit.resolve();
+
+            if (auth !== "password/administrator") {
+                console.log("Installing kerberos keytab not supported for auth mode", auth);
+                return cockpit.resolve();
+            }
+
+            var dfd = cockpit.defer();
+            var user = $(".realms-op-admin").val();
+            var password = $(".realms-op-admin-password").val();
+
+            // ipa-getkeytab needs root to create the file
+            var script = 'set -eu; [ $(id -u) = 0 ] || exit 0; ';
+            // not an IPA setup? cannot handle this
+            script += 'type ipa >/dev/null 2>&1 || exit 0; ';
+
+            // IPA operations require auth; read password from stdin to avoid quoting issues
+            // if kinit fails, we can't handle this setup, exit cleanly
+            script += 'kinit ' + user + '@' + kerberos.RealmName + ' || exit 0; ';
+
+            // create a kerberos Service Principal Name for cockpit-ws, unless already present
+            script += 'service="HTTP/$(hostname -f)@' + kerberos.RealmName + '"; ' +
+                      'ipa service-show "$service" || ipa service-add --ok-as-delegate=true --force "$service"; ';
+
+            // add cockpit-ws key, unless already present
+            script += 'mkdir -p /etc/cockpit; ';
+            script += 'klist -k /etc/cockpit/krb5.keytab | grep -qF "$service" || ' +
+                      'ipa-getkeytab -p HTTP/$(hostname -f) -k /etc/cockpit/krb5.keytab; ';
+
+            // use a temporary keytab to avoid interfering with the system one
+            cockpit.script(script, [], { superuser: "require", err: "message",
+                                         environ: ["KRB5CCNAME=/run/cockpit/keytab-setup"] })
+                .input(password)
+                .done(dfd.resolve)
+                .fail(function(ex) {
+                    dfd.reject(cockpit.message(ex));
```

Comment:
Why is the call to `cockpit.message` necessary here?  Doesn't `cockpit.script` do that already?

Reply:
cockpit.script is just a wrapper around cockpit.spawn, and [process.fail](http://cockpit-project.org/guide/latest/cockpit-spawn.html) gives an exception object with `.message` and/or `.problem` fields. But realm's `perform()` function's error handling expects a plain string, which is why I call cockpit.message here. If it looks weird here, we can also move it into the [caller](https://github.com/cockpit-project/cockpit/pull/8956/files#diff-ca3ecc7c90325f10d3bd683523a2043aR446).

Reply:
I now tried to just throw it as-is, and the error message indeed appears just fine in the dialog (I broke the created service name to trigger a failure). So it seems the dialog already calls `toString()` properly by itself. This simplifies the code to just `.fail(dfd.reject);`, yay! Pushed.

Reply:
LGTM!